### DISCUSSION TITLE
is cel-go good as an FaaS alternative?

### DISCUSSION THREAD

**Discussion by sprappcom** [google/cel-go#967] [Q&A]
with reference here, would like to know if it's advisable to use cel-go as a FaaS alternative like cloudflare workers.
https://github.com/expr-lang/expr/issues/669

**Comment by TristonianJones** [comment]
CEL is non-Turing complete, so it means you can offer pretty tight guarantees on memory and CPU impacts of expressions from otherwise untrusted users without the need for sandboxing. If you're trying to expose a platform features as a function library that's accessible from CEL, then it can be a reasonable way to offer cheap, safe, yet limited scripting capabilities within your platform.

You can also compile in Go and evaluate in CEL in C++ or Java, so depending on your infra requirements you have some flexibility to move CEL logic closer to the data or to a more performant / popular runtime with ease.

**Reply by sprappcom** [reply]
@TristonianJones 1. is it a good "FaaS" alternative? i can write my own mysql / redis connector but not sure if cel-go can be used to interact with them
2. if i use it as FaaS, what are the limitations / best case scenario programs / apps my users can write in?
3. i dont see much caveats / limitations written. does it have a memory use limit?

**Comment by vivshankar** [comment]
@sprappcom FWIW we are using an enhanced version of CEL as a safer method of running untrusted code in a multi tenanted SaaS environment. We have a number of additional custom types added but we have tried to ensure we don't inadvertently introduce tight loops etc. 

I wrote this article a couple years back providing a high level view. It might be of interest.

https://medium.com/@viv.shankar/running-untrusted-code-in-saas-d27512bd94d1

Regarding your questions -

1. You can write custom types for http clients, redis clients etc but I would generally recommend wrapping it in some form of business objects as opposed to providing the raw functions. "Caches" for example that enforce a ttl, have loader functions that run with restrictions, introduce telemetry etc, as opposed to a redis get and set. 
2. I think this will be based on your use cases. You will run into limitations in terms of utilities. Ultimately, CEL is single line expressions. If you need if blocks, assignments etc, you will end up implementing them. We did it as yaml, rather than write a full fledged programming language. 
3. This would be dependent on your custom adapters and compute environment. 

There are other factors IMO that will determine whether you would want to use it. Answering some of these helped me reach a decision.

1. Would your user base accept this and be willing to learn? Mainstream languages like JavaScript and python are obviously more familiar to people. 
2. Are you looking at building a lambda type service or something that is meant for power users as part of a larger set of features?
3. What sort of connectors and functions would you want to build? Do you have the appetite to build them?
